src/compiler/crystal/zero_one_or_many.cr

```diff
@@ -0,0 +1,110 @@
+# An Array(T)-like type that's optimized for the case of
+# frequently having zero or one elements.
+#
+# It's a struct, so if you have a ZeroOneOrMany and you
+# want to mutate, you have to use `with` or `without`
+# and overwrite the value you have with what you get
+# from those methods.
+#
+# Warning: you should never keep the original value
+# around when using the above methods as the internal
+# array might be mutated (for performance reasons),
+# so the original value might mutate in that case!
+#
+# TODO: add tests for this type.
+struct Crystal::ZeroOneOrMany(T)
+  include Enumerable(T)
+
+  getter value : Nil | T | Array(T)
+
+  def initialize
+    @value = nil
+  end
+
+  def initialize(@value : T)
+  end
+
+  def initialize(values : Array(T))
+    @value =
+      case values.size
+      when 0
+        nil
+      when 1
+        values.first
+      else
+        values
+      end
+  end
```

Comment:
> Could we just store the array, regardless of how many elements it has?

If we do that, and this `ZeroOneOrMany` is stored inside an ASTNode's dependencies or observers, and that node isn't GCed at all until the end of the compilation, it means that the memory for the node and the array must be kept at all time in memory. If we only keep the single element the array's memory can be GCed.

Does that make sense?

>  re-using that array

For this we should document that the array passed here should not be reused for something else. The owner if the array is now `ZeroOneOrMany`. I consider this type a bit low-level. That's why it might not be a good idea to have it in the standard library.